### ISSUE TITLE
Unable reproduce results in the Wizard of Wikipedia paper

### ISSUE THREAD

**Issue by a2jain** [facebookresearch/ParlAI#1843]
We are training the **TwoStageAgent generator** from scratch to reproduce Gold Knowledge results similar to those of Table 4 (Generative models on the full Wizard Task). We are using the **default parameters** for training and changing the two parameters as mentioned in the paper (**beam size = 5, bpe encoding**).

We are wondering if we should be training with a certain set of parameters different than the default ones to get similar results. With the current set of parameters, we are not able to produce the paper’s Test Seen: 30.0 and Test Unseen: 42.7. Instead, we are getting high perplexities with a difference of more than 20 points.

**Comment by stephenroller** [comment]
Thanks for moving to GitHub. It helps create a paper trail we can refer people to if the question comes up again!

You're trying to produce the one with knowledge dropout or the one without? The code we released is a re-implementation of the paper, so it might not get exactly the same, but we should be able to get pretty close.

These are the hyperparameters I used for the best 2 stage model without knowledge dropout, and they're quite different from the defaults I think.

```
--validation-every-n-secs 120  # probably -veps 0.25 is more stable
--validation-metric ppl
--validation-metric-mode min
--save-after-valid True
--load-from-checkpoint True
--dict-lower True
--dict-maxexs -1
--dict-tokenizer bpe
--dict-include-valid False
--log-every-n-secs 30
--fp16 True
--lr 0.0005
--ffn-size 512
--clip-norm 0.1
--optimizer adam
--lr-scheduler inverse_sqrt
--warmup-updates 5000
--embedding-type fasttext
-esz 256
--n-heads 2
--n-layers 3
--dropout 0.25
-bs 64
```

Could you try those? If those still get far from off, then I'll need to try to recreate myself and see if something broke (ParlAI is a living code base).

**Comment by a2jain** [comment]
Thank you, Stephen, for a prompt reply. I will try with this set of parameters and will update about its status.

**Comment by stephenroller** [comment]
jfyi it's `--gradient-clip` instead of `--clip-norm`

**Comment by a2jain** [comment]
Should I be expecting similar results if the `--fp16` is set to `False` instead of `True`? As fp16 is for speedup gains.

**Comment by a2jain** [comment]
I tried with `--fp16 False` and got **Test Seen PPL: 32.25** and **Test Unseen PPL: 48.92**. These values are close to the ones in paper, although still on the higher side. But I guess, this works. Correct me if you think they should have been even lower.

**Comment by stephenroller** [comment]
We're in the right ball park, but they probably a bit closer. I would've expected ~31 and ~45 should've been pretty easy to create. There might be some small differences in the vocabulary or some details of the reimplementation (There was some weirdness in positional embeddings before). That unseen ppl is what I'm most concerned about (42 and 49 are quite far apart...).

Correct, fp16 is just a speed trick. However, if you had to lower the batchsize to handle the increased memory usage, then maybe that would affect things; if you used `--validation-every-n-sec 120` with fp32, then it could be *easily* early stopping too early. In my old logs, training took ~6000 seconds and finally stopped after 43.5 epochs. With fp32 I'd expect ~5 hours on a single GPU. Training ppl was ~15.7. The model finished with 29.48 validation seen ppl.

How many epochs did your model go through?

**Comment by stephenroller** [comment]
Also, it looks like with my validate-every-120-seconds, it was actually validating close to every ~1 epoch, and the actual best-model was at ~28 epochs.

**Comment by a2jain** [comment]
So, I ran with `--validation-every-n-sec 120` and `--fp16 False`. The training did not actually stop, but it went till 55 epochs with impatience of 6. And the best validation seen ppl was 31.39. Should I not stop the training at all?

When you say that the best-model was at ~28 epochs, do you mean the validation seen ppl was the lowest at around that epoch?

**Comment by stephenroller** [comment]
Yes, it began overfitting at 28 epochs.

Hopefully this model you have is enough for you to continue your own work. Sometime this week or next, I’ll spend some effort seeing if I can close the gap.

Closing this issue for now, but please reopen if you feel blocked and have more questions.

**Comment by stephenroller** [comment]
My recreation with the same parameters above, except `-veps 1.0 --fp16 true` ended up being 31.85/46.87, so about the same as yours, but a bit closer.

My hunch is that there is a bug with how BPE is handling the special tokens in this reimplementation (I suspect it's breaking them into subtokens instead of keeping them as single items). That would cause truncation to be a few tokens off, and might explain the rest of the difference.

**Comment by a2jain** [comment]
Ok, these results are good enough for me to carry on with my work. Thanks for your help! Also, just wondering if something should be done to test your hunch on the BPE bug.

**Comment by stephenroller** [comment]
I confirmed that the current model is breaking the tokens up. It’s costing about 6 extra tokens compared to what the original model was doing. It’s probably responsible for the rest of the delta.

Since you’re unblocked I probably won’t put any more time into this until end of the week. It’s a pretty easy to hack it in to test this hypothesis, but might require discussion with my colleagues to make sure this doesn’t happen again.

**Comment by acha21** [comment]
This is maybe stupid question. How can we train the two stage model with gold or predicted knowledge.
I ran with

**Comment by stephenroller** [comment]
What you have above is the gold knowledge model. It becomes two stage when you put another (ranking) model in front of it that actually choses the knowledge.

We didn't release the ranker for the first-stage of the two-stage model, and I'm not sure that we consider it a priority at the moment. Frankly we anticipated that the end2end models would be more interesting to the community, and there was a lot of really dirty glue that was used to stitch together the two-stage model. We can revisit this decision if it becomes very crucial to the community.

**Comment by sshleifer** [comment]
Might someone check my wildly speculative understanding of the current code?

`projects/wizard_of_wikipedia/generator/train_end2end.py`:

1.  starts without any reddit/squad pretraining.
2.  corresponds to the last two rows of Table 4 in the paper. Which row depends on knowledge_alpha=0.95, which controls knowledge dropout.
3.  Does not use gold knowledge, uses "predicted" knowledge.
4.  is not tested in the human eval experiments (Table 5), but a separate generative model that doesn't use any knowledge is.
5.  At chat time, does this model evaluates a large number of candidate knowledge sentences, determined by the IR system.
6.  How long did it take to train?

### Other Models
7. The mechanism for training a two stage model that uses gold knowledge (knows `chosen_sentence`) involves using the parser options discussed in this thread, and trains the second stage `TwoStageAgent`, there is no easy way to train the first stage of`TwoStageAgent` at the moment.
8. Are any of the systems evaluated in Table 5 that are associated with a training script/easy to train with the current code?

Thanks in advance!

**Comment by stephenroller** [comment]
- Knowledge alpha is the loss interpolator from the paper. `--ignorant-dropout` is the "knowledge dropout". (Bad name, I admit).
- IIRC it took something like 6 hours to train on a single v100, but I could be off.
- I'm actually not sure if the released retrieval model has an easy to train script. There were some issues with releasing the pretrained model at some point, but I don't know where that stands now. @emilydinan?

I believe the rest of your statements are correct.
